lib/rspec/mocks/targets.rb

```diff
@@ -38,7 +38,9 @@ def self.disallow_negation(method_name)
     private
 
       def matcher_allowed?(matcher)
-        matcher.class.name.start_with?("RSpec::Mocks::Matchers".freeze)
+        ancestors = matcher.class.ancestors
+        namespace = "RSpec::Mocks::Matchers"
+        ancestors.any? { |klass| klass.name.to_s.start_with? namespace }
```

Comment:
Suggestion: instead of doing a linear search over all the ancestors and _implicitly_ using the matcher class name, what do you think of creating an explicit "tag" module that you can check instead?  I'm thinking of something like this:

``` ruby
module RSpec
  module Mocks
    module Matchers
      # exists just so we can "tag" all rspec-mocks matchers to make detection of them easier.
      module Matcher; end

      # now include the module in each of our matcher classes...
      class HaveReceived
        include Matcher
      end

      class Receive
        include Matcher
      end

      class ReceiveMessages
        include Matcher
      end

      class ReceiveMessageChain
        include Matcher
      end

      # ...and then we can detect rspec-mocks matchers with a simple `===` check:
      def matcher_allowed?(matcher)
        RSpec::Mocks::Matchers::Matcher === matcher
      end
    end
  end
end
```